lib/Sema/CSApply.cpp

```diff
@@ -6818,8 +6818,9 @@ ExprRewriter::finishApplyDynamicCallable(const Solution &solution,
       cs.setType(labelExpr, keyType);
       handleStringLiteralExpr(cast<LiteralExpr>(labelExpr));
 
-      Expr *pair =
-        TupleExpr::createImplicit(ctx, { labelExpr, arg->getElement(i) }, {});
+      Expr *pair = TupleExpr::createImplicit(
+          ctx, {labelExpr, coerceToType(arg->getElement(i), valueType, loc)},
```

Comment:
Unfortunately `coerceToType` can fail, which would lead to an invalid AST for TupleExpr. I suggest you do this coercion separately and return `nullptr` if it fails, that way at the very list we are going to produce a fallback diagnostic.

Reply:
We're slowly trying to eliminate all the ways in which it can fail, though. I suspect most of the remaining ones are due to an invalid stdlib (missing literal protocols, etc) which can just become assertions.